File: huffman.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
using namespace std;

// Node structure
struct Node {
    char ch;
    int freq;
    Node *left, *right;

    Node(char c, int f) {
        ch = c;
        freq = f;
        left = right = nullptr;
    }
};

// Compare structure for min-heap
struct Compare {
    bool operator()(Node* a, Node* b) {
        return a->freq > b->freq;
    }
};

// Generate Huffman codes
void generateCodes(Node* root, string code, unordered_map<char, string>& codes) {
    if (!root)
        return;
    if (root->ch != '\0')
        codes[root->ch] = code;

    generateCodes(root->left, code + "0", codes);
    generateCodes(root->right, code + "1", codes);
}

// Build Huffman Tree
Node* buildHuffmanTree(unordered_map<char, int>& freq) {
    priority_queue<Node*, vector<Node*>, Compare> pq;

    for (auto pair : freq)
        pq.push(new Node(pair.first, pair.second));

    while (pq.size() > 1) {
        Node* left = pq.top(); pq.pop();
        Node* right = pq.top(); pq.pop();

        Node* merged = new Node('\0', left->freq + right->freq);
        merged->left = left;
        merged->right = right;

        pq.push(merged);
    }

    return pq.top();
}

// ---- Main Program ----
int main() {
    string text;
    cout << "Enter a string: ";
    getline(cin, text);

    unordered_map<char, int> freq;
    for (char ch : text)
        freq[ch]++;

    Node* root = buildHuffmanTree(freq);

    unordered_map<char, string> codes;
    generateCodes(root, "", codes);

    cout << "\nHuffman Codes:\n";
    for (auto pair : codes)
        cout << pair.first << ": " << pair.second << endl;

    cout << "\nEncoded Text: ";
    for (char ch : text)
        cout << codes[ch];
    cout << endl;

    return 0;
}

File: huffman.py
import heapq

# Node class for Huffman Tree
class Node:
    def __init__(self, char, freq):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    # Comparison operator for min-heap
    def __lt__(self, other):
        return self.freq < other.freq


# Function to build Huffman Tree
def build_huffman_tree(char_freq):
    heap = [Node(char, freq) for char, freq in char_freq.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)

        merged = Node(None, left.freq + right.freq)
        merged.left = left
        merged.right = right

        heapq.heappush(heap, merged)

    return heap[0]  # root node


# Generate Huffman Codes
def generate_codes(root, code="", codes={}):
    if root is None:
        return

    if root.char is not None:
        codes[root.char] = code
    generate_codes(root.left, code + "0", codes)
    generate_codes(root.right, code + "1", codes)
    return codes


# ---- Main Program ----
text = input("Enter a string: ")

# Calculate frequency of each character
freq = {}
for ch in text:
    freq[ch] = freq.get(ch, 0) + 1

# Build Huffman Tree
root = build_huffman_tree(freq)

# Generate Huffman Codes
codes = generate_codes(root)

# Print codes
print("\nHuffman Codes:")
for ch in codes:
    print(f"{ch}: {codes[ch]}")

# Encode input text
encoded_text = ''.join(codes[ch] for ch in text)
print("\nEncoded Text:", encoded_text)
